drafts/str4d-stratum/draft1.rst

```diff
@@ -0,0 +1,256 @@
+Abstract
+========
+
+This ZIP describes the Zcash variant of the Stratum protocol, used by miners to
+communicate with mining pool servers.
+
+
+Specification
+=============
+
+``mining.subscribe()``
+----------------------
+
+Request::
+
+    {"id": 1, "method": "mining.subscribe", "params": ["CONNECT_HOST", CONNECT_PORT, "MINER_USER_AGENT", "SESSION_ID"]}\n
+
+- ``CONNECT_HOST`` (str): The host string used by the miner.
+
+  - ``pool.example.com``
+
+- ``CONNECT_PORT`` (int): The port used by the miner.
+
+  - ``3337``
+
+- ``MINER_USER_AGENT`` (str): A free-form string specifying the type and version
+  of the mining software. Recommended syntax is the User Agent format used by
+  Zcash nodes.
+
+  - ``Zatoshi/1.0.0``
+
+- ``SESSION_ID`` (str): The id for a previous session that the miner wants to
+  resume (e.g. after a temporary network disconnection).
+
+  - MAY be ``null`` indicating that the miner wants to start a new session.
+
+Response::
+
+    {"id": 1, "result": ["NONCE_1", "SESSION_ID"], "error": null}\n
+
+- ``SESSION_ID`` (str): The session id, for use when resuming.
+
+  - MAY be ``null`` indicating that the server does not support session
+    resuming.
```

Comment:
> why does the protocol need to be stateful anyway?

General idea (not mine) is that miner can mine even on disconnection and then submit everything on reconnect. Submits may be valid if there's no new block in meantime. I personally think this is over-optimization and I'm okay with removing session parameter from the protocol entirely.